nats-base-client/headers.ts

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2020 The NATS Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Heavily inspired by Golang's https://golang.org/src/net/http/header.go
+
+import { ErrorCode, NatsError } from "./error.ts";
+
+export interface Headers {
+  get(k: string): string;
+  set(k: string, v: string): void;
+  append(k: string, v: string): void;
+  has(k: string): boolean;
+  values(k: string): string[];
+  delete(k: string): void;
+}
+
+export class MsgHeaders implements Headers {
+  static HEADER = "NATS/1.0";
+  error?: number;
+  headers: Map<string, string[]> = new Map();
+
+  constructor() {}
+
+  size(): number {
+    let count = 0;
+    for (const [_, v] of this.headers.entries()) {
+      count += v.length;
+    }
+    return count;
+  }
+
+  equals(mh: MsgHeaders) {
+    if (
+      mh && this.headers.size === mh.headers.size &&
+      this.error === mh.error
+    ) {
+      for (const [k, v] of this.headers) {
+        const a = mh.values(k);
+        if (v.length !== a?.length) {
+          return false;
+        }
+        const vv = [...v].sort();
+        const aa = [...a].sort();
+        for (let i = 0; i < vv.length; i++) {
+          if (vv[i] !== aa[i]) {
+            return false;
+          }
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+  static decode(a: Uint8Array): MsgHeaders {
+    const mh = new MsgHeaders();
+    const s = new TextDecoder().decode(a);
+    const lines = s.split("\r\n");
+    const h = lines[0];
+    if (h !== MsgHeaders.HEADER) {
+      const str = h.replace(MsgHeaders.HEADER, "");
+      mh.error = parseInt(str, 10);
```

Comment:
Not sure I understand this code, but are you parsing the status if it is sent like this: `NATS/1.0 503\r\n` here? If not, know that it will be coming in another feature (the no responders) where we may get this kind of header).

Reply:
yes this is wip - next commit will answer it - as it will fail the request with an error.

Reply:
https://github.com/nats-io/nats.deno/pull/8/files#diff-812a477358fb5cc1c60403e0b9fe3b35R301 is where this is looked at.